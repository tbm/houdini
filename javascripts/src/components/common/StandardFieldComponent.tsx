// License: LGPL-3.0-or-later
import * as React from 'react';

export interface StandardFieldComponentProps
{
    inError:boolean
    error?:string
    children?:React.ReactNode
    [additional_properties:string]: any
}

export default class StandardFieldComponent extends React.Component<StandardFieldComponentProps, {}> {
    constructor(props:StandardFieldComponentProps){
        super(props)
    }
    renderChildren(childStyle:React.CSSProperties){
        return React.Children.map(this.props.children, child  => {
            return React.cloneElement(child as React.ReactElement<any>,  {
              style: childStyle
            })
        })
    }
  render() {
    let errorState = this.props.inError ? 'invalid' : 'valid'
    let errorMessage = this.props.inError ? this.props.error : undefined
    let styles = this.props.inError ? {borderColor: "#FF4F4F"} : {}
    let errorDivStyle = this.props.inError ? {display:"block"} : {display:"none"}
    let errorDiv = <div style={errorDivStyle} role="alert">{errorMessage}</div>

    return <>
      <div>
        {this.renderChildren(styles)}
        {errorDiv}
      </div>

      </>
  }
}

