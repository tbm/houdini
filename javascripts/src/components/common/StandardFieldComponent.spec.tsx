// License: LGPL-3.0-or-later
import * as React from 'react';
import 'jest';
import {shallow,render} from 'enzyme'
const TestRenderer = require('react-test-renderer')
import StandardFieldComponent from './StandardFieldComponent'

describe('StandardFieldComponent', () => {
    test('works with no children', () => {
        var field = TestRenderer.create((<StandardFieldComponent inError={false} />)).toJSON()


        expect(field).toMatchSnapshot()
    })
    test('works with a child', () => {
        var field = TestRenderer.create(<StandardFieldComponent inError={false}><input/></StandardFieldComponent>);

        expect(field).toMatchSnapshot()
    })

    test('sets error message properly', () => {
        var field = TestRenderer.create(<StandardFieldComponent inError={true} error={"Something more"}><input/></StandardFieldComponent>);

        expect(field).toMatchSnapshot()
    })
})