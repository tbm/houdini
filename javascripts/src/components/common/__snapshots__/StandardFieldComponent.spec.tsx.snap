// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardFieldComponent sets error message properly 1`] = `
<div
  data-ff-field="invalid"
  data-ff-field-error="Something more"
>
  <input
    data-ff-field-input="invalid"
  />
</div>
`;

exports[`StandardFieldComponent works with a child 1`] = `
<div
  data-ff-field="valid"
  data-ff-field-error=""
>
  <input
    data-ff-field-input="valid"
  />
</div>
`;

exports[`StandardFieldComponent works with no children 1`] = `
<div
  data-ff-field="valid"
  data-ff-field-error=""
/>
`;
